src/arithmetic/agg_funcs.c

```diff
@@ -206,17 +301,41 @@ int __agg_countStep(AggCtx *ctx, SIValue *argv, int argc) {
 	return AGG_OK;
 }
 
+int __agg_countDistinctStep(AggCtx *ctx, SIValue *argv, int argc) {
+	__agg_countCtx *ac = Agg_FuncCtx(ctx);
+	// Batch size to this function is always one, so
+	// we only need to check the first argument
+	SIValue value = *argv;
+	if(!SIValue_IsNullPtr(argv) && Set_Add(ac->hashSet, *argv))
+		ac->count ++;
+	return AGG_OK;
+}
+
 int __agg_countReduceNext(AggCtx *ctx) {
 	__agg_countCtx *ac = Agg_FuncCtx(ctx);
 	Agg_SetResult(ctx, SI_LongVal(ac->count));
 	return AGG_OK;
 }
 
-AggCtx *Agg_CountFunc() {
-	__agg_countCtx *ac = malloc(sizeof(__agg_countCtx));
+void *__agg_countCtxNew() {
+	__agg_countCtx *ac = rm_malloc(sizeof(__agg_countCtx));
 	ac->count = 0;
+	ac->hashSet = Set_New();
+	return ac;
+}
+
+void __agg_countCtxFree(AggCtx *ctx) {
+	__agg_countCtx *ac = Agg_FuncCtx(ctx);
+	Set_Free(ac->hashSet);
+	rm_free(ac);
+}
 
-	return Agg_Reduce(ac, __agg_countStep, __agg_countReduceNext);
+AggCtx *Agg_CountFunc(bool distinct) {
+	__agg_countCtx *ac = __agg_countCtxNew();
+	if(distinct) return Agg_Reduce(ac, __agg_countDistinctStep, __agg_countReduceNext,
+									   __agg_countCtxNew, __agg_countCtxFree);
+	else return Agg_Reduce(ac, __agg_countStep, __agg_countReduceNext, __agg_countCtxNew,
+							   __agg_countCtxFree);
```

Comment:
My personal preference is to use curly braces when `if` and `else` can't be on contiguous lines, but I'm willing to defer to your taste on this.